extensions/data-workspace/src/services/workspaceService.ts

```diff
@@ -158,6 +159,61 @@ export class WorkspaceService implements IWorkspaceService {
 		return vscode.workspace.workspaceFile ? this.getWorkspaceConfigurationValue<string[]>(ProjectsConfigurationName).map(project => this.toUri(project)) : [];
 	}
 
+	/**
+	 * Check for projects that are in the workspace folders but have not been added to the workspace through the dialog or by editing the .code-workspace file
+	 */
+	async checkForProjectsNotAddedToWorkspace(): Promise<void> {
+		const config = vscode.workspace.getConfiguration(constants.projectsConfigurationKey);
+
+		// only check if the user hasn't selected not to show this prompt again
+		if (config[constants.showNotAddedProjectsMessageKey]) {
+			// look for any projects that haven't been added to the workspace
+			const projectsInWorkspace = this.getProjectsInWorkspace();
+			const workspaceFolders = vscode.workspace.workspaceFolders;
+
+			if (workspaceFolders) {
```

Comment:
I think it's generally better to have early escapes rather than indenting the bulk of logic so everything's more readable (logic doesn't wrap as quickly, you don't have to scroll far to the bottom just to find there's no "else")

```
// only check if the user hasn't selected not to show this prompt again
if (!config[constants.showNotAddedProjectsMessageKey]) {
    return;
}

[...]

if (!workspaceFolders) {
    return;
}

[...]
```